pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,481 @@
+package networkpolicy
+
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"net"
+	"sync"
+	"time"
+
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	"antrea.io/antrea/pkg/agent/types"
+	"antrea.io/antrea/pkg/agent/util"
+	crdv1beta "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+	"antrea.io/antrea/pkg/util/channel"
+)
+
+type ruleType int
+type eventType uint8
+
+const (
+	unicast   ruleType = 0
+	igmp      ruleType = 1
+	multicast ruleType = 2
+
+	groupJoin   eventType = 0
+	groupLeave  eventType = 1
+	rulechanged eventType = 2
+)
+
+var (
+	mcastAllHosts           = net.ParseIP("224.0.0.1").To4()
+	_, mcastAllHostsCIDR, _ = net.ParseCIDR("224.0.0.1/32")
+)
+
+type mcastItem struct {
+	groupAddress net.IPNet
+	ruleIDs      map[string]*uint16
+}
+
+type GroupMemberStatus struct {
+	group net.IP
+	// localMembers is a map for the local Pod member and its last update time, key is the Pod's interface name,
+	// and value is its last update time.
+	localMembers   map[string]time.Time
+	lastIGMPReport time.Time
+	mutex          sync.RWMutex
+	ofGroupID      binding.GroupIDType
+}
+
+type mcastGroupEvent struct {
+	eType eventType
+	iface *interfacestore.InterfaceConfig
+}
+
+type multicastController struct {
+	ofClient openflow.Client
+
+	ifaceStore interfacestore.InterfaceStore
+
+	queryGroupId     binding.GroupIDType
+	groupInstalled   bool
+	queryGroupStatus GroupMemberStatus
+
+	ruleCache             *ruleCache
+	mcastItemRuleIDMap    map[string]mcastItem
```

Comment:
Add comments to explain the key and value of this map, so does ruleIDGroupAddressMap

Reply:
done